Chore switch to postgres db

Cheza-Dzabala opened the pull request:
#### What does this PR do?

Removes all data structures from the application & implements a Postgresql Database.

#### Description of Task to be completed?

Removing all data structures from the application.  Replaced the datas tructures with 3 data bases:
 
Users
 ------------------ 
**id** | SERIAL INTEGER PRIMARY KEY
**first_name** | VARCHAR (128) NOT NULL
**last_name** | VARCHAR (128) NOT NULL
**email** | VARCHAR (128) NOT NULL
**password** | VARCHAR (128) NOT NULL


 Trips
 ------------------ 
**id** | SERIAL INTEGER PRIMARY KEY
**trip_date** | VARCHAR (128) NOT NULL
**bus_licesnse_number** | VARCHAR (128) NOT NULL
**origin** | VARCHAR (128) NOT NULL
**destination** | VARCHAR (128) NOT NULL
**fare** | VARCHAR (128) NOT NULL
**seating_capacity** | INTEGER NOT NULL
**number_of_seats** | INTEGENOT NULL


 Bookings
 ------------------ 
**id** | SERIAL INTEGER
**seat_number** | VARCHAR (128) NOT NULL
**user_id** | VARCHAR (128) NOT NULL
**trip_id** | VARCHAR (128) NOT NULL
PRIMARY KEY (user_id, trip_id)

#### How should this be manually tested?

1. Install PostgreSQL by following [these](https://www.postgresql.org/docs/9.3/tutorial-install.html) instructions.
2. Start PostgreSQL. 
3. Clone the repo. 
4. Install all dependencies by running ` npm install `.
5. Run the command: ` npm start `
6. Assuming you're on localhost:3000, sign up for account at ` localhost:3000/api/v2/auth/signup `
7. Check PostgreSQL for persistance. 
<-- opttional -->

To access protected routes:

1. Default user email: ` dzabalamacheza@gmail.com ` & password ` Runfree8418_!* `
2. To create your own admin, access route:
POST ` localhost:3000/api/v2/admins `

###### Entity Specification
```javascript
body: {
  "email": String,
   "first_name": String,
    "last_name": String,
    "password": String
}
````

###### Response Specification
```javascript
body: {
  "status": "success",
 "id": Int,
   "first_name": String,
    "last_name": String,
}
````

#### Any background context you want to provide?

Data was previously stored in Arrays

#### What are the relevant pivotal tracker stories?

###### #167826363
#### Screenshots (if appropriate)

<img width="1447" alt="Screenshot 2019-08-14 at 9 09 09 AM" src="https://user-images.githubusercontent.com/13886438/63001377-ceb51080-be73-11e9-925a-41eaccea11ca.png">

![screencapture-127-0-0-1-49164-browser-2019-08-14-09_09_20](https://user-images.githubusercontent.com/13886438/63001370-cc52b680-be73-11e9-9fd3-23be53f40d3f.png)


<img width="1460" alt="Screenshot 2019-08-14 at 9 08 32 AM" src="https://user-images.githubusercontent.com/13886438/63001378-cf4da700-be73-11e9-86a4-53a99de70d72.png">


<img width="1817" alt="Screenshot 2019-08-14 at 9 18 17 AM" src="https://user-images.githubusercontent.com/13886438/63002072-62d3a780-be75-11e9-90c3-3f8390fa3c56.png">

<img width="1461" alt="Screenshot 2019-08-14 at 9 07 41 AM" src="https://user-images.githubusercontent.com/13886438/63001379-cfe63d80-be73-11e9-9672-4b746bd26ecd.png">

![screencapture-127-0-0-1-49164-browser-2019-08-14-09_07_49](https://user-images.githubusercontent.com/13886438/63001373-cd83e380-be73-11e9-8898-e5c0f78d4ad9.png)

Updated Codebase
-------------------

<img width="1200" alt="Screenshot 2019-08-14 at 7 48 50 PM" src="https://user-images.githubusercontent.com/13886438/63043597-d654d380-becc-11e9-9339-894844adc22f.png">
<img width="985" alt="Screenshot 2019-08-14 at 7 48 37 PM" src="https://user-images.githubusercontent.com/13886438/63043599-d654d380-becc-11e9-986b-2df5db3978ba.png">
<img width="666" alt="Screenshot 2019-08-14 at 7 48 18 PM" src="https://user-images.githubusercontent.com/13886438/63043601-d654d380-becc-11e9-82e4-454145b6e81d.png">
<img width="537" alt="Screenshot 2019-08-14 at 7 48 06 PM" src="https://user-images.githubusercontent.com/13886438/63043602-d6ed6a00-becc-11e9-85ce-e2fa5501822d.png">
<img width="841" alt="Screenshot 2019-08-14 at 7 47 53 PM" src="https://user-images.githubusercontent.com/13886438/63043603-d6ed6a00-becc-11e9-8b3b-8d6da85bb9e6.png">
<img width="728" alt="Screenshot 2019-08-14 at 7 47 34 PM" src="https://user-images.githubusercontent.com/13886438/63043604-d6ed6a00-becc-11e9-8e2b-51d8e365162f.png">
<img width="527" alt="Screenshot 2019-08-14 at 7 47 24 PM" src="https://user-images.githubusercontent.com/13886438/63043605-d7860080-becc-11e9-870c-b9caf6c75baf.png">

#### Questions:

None

Cheza-Dzabala committed:
Chore replace data structures with postgresql

- Removed all data structures

- Created Postgres Database for Way farer

- Minor refactoring done to test folder

[starts #167826363]

Cheza-Dzabala committed:
Chore replace data structures with postgresql

- Removed all data structures

- Created Postgres Database for Way farer

- Minor refactoring done to test folder

[starts #167826363]

Cheza-Dzabala committed:
Merge branch 'ch-switch-to-postgres-db' of https://github.com/Cheza-Dzabala/way-farer into ch-switch-to-postgres-db

Cheza-Dzabala committed:
Chore setup database

- Added postgres SU for Travis CI Build on the .travis.yml file

[167826363]

Cheza-Dzabala committed:
Chore setup database

- Rmoved custom user for Travis CI Build on the .travis.yml file

[167826363]

Oluwasegun-AA commented on server/v2/app.js:
```diff
@@ -0,0 +1,35 @@
+app.use(`/api/${version}/auth`, authRoutes);
+app.use(`/api/${version}/trips`, tripRoutes);
+app.use(`/api/${version}/bookings`, bookingRoutes);
+app.use(`/api/${version}/admins`, adminRoutes);
```
@Cheza-Dzabala , with respect to the directions IU gave regarding making your application modular, you should only use a single middleware for your custom routes.  this should be imported from the index.js file in the middleware folder

Cheza-Dzabala replied:
Well noted.

Oluwasegun-AA commented on server/v2/helpers/bookingHelpers/bookingRelationships.js:
```diff
@@ -0,0 +1,17 @@
+  req.body.trip_id = trip.id;
+  req.body.email = user.email;
+  return next();
+};
```
@Cheza-Dzabala , this is a middleware, it should not be in the helpers folder; rather this should be in the middleware folder

Cheza-Dzabala replied:
Noted.. Will resolve this immidiately.

Oluwasegun-AA commented on server/v2/helpers/bookingHelpers/bookingSeatChecker.js:
```diff
@@ -0,0 +1,19 @@
+  if (flag.status === true) return Response(res, 409, 'unsuccessful', { message: flag.message });
+
+  return next();
+};
```
@Cheza-Dzabala , middlwares should not be in the helpers folder

Cheza-Dzabala replied:
Noted

Oluwasegun-AA commented on server/v2/helpers/bookingHelpers/verifyBooking.js:
```diff
@@ -0,0 +1,9 @@
+  const booking = await bookingsModel.findBooking(id);
+  if (!booking) return Response(res, 404, 'unsuccessful', { message: 'Booking already deleted' });
+  return next();
+};
```
@Cheza-Dzabala , i can see you have so many more middlewares in the helpers folder. please remove them and structure your files accordingly

Cheza-Dzabala replied:
Noted.

Oluwasegun-AA commented on server/v2/helpers/tripHelpers/tripsRelationships.js:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  return next();
+};
```
@Cheza-Dzabala please explain in this conversation what you are trying to achieve here

Oluwasegun-AA commented on server/v2/middleware/userTypeMiddleware.js:
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable dot-notation */
```
@Cheza-Dzabala , you are required the follow strictly the Airbnb Eslint configuration. in lieu of this, you are not expected to bypass an enforced style using the Esling comments. please remove all comments and ensure your code conforms to the Airbnb standard without errors

Cheza-Dzabala replied:
Yes, I found the red underlining to be annoying but I will work with it and refactor to the recommended language

Oluwasegun-AA commented on server/v2/middleware/validationMiddleware/signupValidationMiddleware.js:
```diff
@@ -0,0 +1,13 @@
+import userModel from '../../models/userModel';
+
+import authValidations from '../../validations/authValidations';
+
```
@Cheza-Dzabala , you might want to remove unnecessary spaces between your import statements

Cheza-Dzabala replied:
WIll do that

Oluwasegun-AA commented on server/v2/models/userModel.js:
```diff
@@ -0,0 +1,76 @@
+async function findUserById(id) {
+  const user = await dbService.findUserById(id);
+  return user;
+}
```
@Cheza-Dzabala , you have adopted a nice apprach. however, I can see you'r repeating methods. why can you have a single class that will handle all kinds of request, whether user/booking or trip?

above there you had a method to fetch `all trips/trips by Id` and another method to fetch `all booking/ booking by Id`. you can basically have just a single method(in the helper file) which implement this `ONCE` for all the tables. hence you call it in all controllers irrespective

Cheza-Dzabala replied:
This is an interesting suggestion. I will adopt it.

Oluwasegun-AA commented on server/v2/package.json:
```diff
@@ -0,0 +1,45 @@
+    "eslint-plugin-import": "^2.18.2",
+    "mocha": "^6.2.0",
+    "mocha-lcov-reporter": "^1.3.0",
+    "nyc": "^14.1.1"
```
@Cheza-Dzabala , please remove all the caret (^)symbols in your dependencies and dev-dependencies. it ensures your files are locked and breaking changes in future version will not affect your workflow even when you rerun this app in the year 2050

Cheza-Dzabala replied:
Noted

Oluwasegun-AA commented on server/v2/services/authDBServices.js:
```diff
@@ -0,0 +1,81 @@
+}
+
+async function signin(email, password) {
+  const queryText = 'SELECT * From users WHERE email = $1';
```
@Cheza-Dzabala , for modularity sake, you can have a helper file which is going to house all your query texts.. so you do somthing like 
```js
import * as query from './queryHelperFile'

const queryText = query.getBookings;

```

Cheza-Dzabala replied:
Ah yes, I will do that.

Oluwasegun-AA commented on server/v2/controllers/signinController.js:
```diff
@@ -0,0 +1,17 @@
+  const { email, password } = body;
+  const user = await UserModel.signin(email, password);
+  if (user) {
+    return authenticationCheck.signNewToken(user, res, 200);
```
@Cheza-Dzabala , are you encrypting all user data here? if yes, that is a bad practice. you should only encrypt few data that will be useful on the front end of things (i.e for making a return request)

Cheza-Dzabala replied:
I have a middleware that encrypts the data. I am sending that middleware the user object, the response and the appropriate status code to return on success.

Cheza-Dzabala commented on server/v2/helpers/tripHelpers/tripsRelationships.js:
```diff
@@ -0,0 +1,20 @@
+  const { body } = req;
+  const busses = await tripModel.findBus(body.bus_license_number);
+  let len = 0;
+  if (busses.length > 0) {
```
Here I determine if busses have been retrieved from the database

Cheza-Dzabala commented on server/v2/helpers/tripHelpers/tripsRelationships.js:
```diff
@@ -0,0 +1,20 @@
+  const busses = await tripModel.findBus(body.bus_license_number);
+  let len = 0;
+  if (busses.length > 0) {
+    while (len < busses.length) {
```
I loop through the busses that were retrieved from the database to determine if they have the same trip date as the trip date in the request body. If the dates are the same, Then I determine that the bus is booked.

Cheza-Dzabala committed:
Chore applied LFA feedback

Modularized Routes

- Added helper functions to clean code

- Improved tests

[#167826363]

Cheza-Dzabala committed:
Chore applied LFA feedback

Modularized Routes

- Added helper functions to clean code

- Improved tests

[#167826363]
